packages/context/test/unit/binding.ts

```diff
@@ -73,40 +74,104 @@ describe('Binding', () => {
       expect(binding.getValue(ctx)).to.equal('value');
     });
+
+    it('sets the style to CONSTANT', () => {
+      binding.to('value');
+      expect(binding.style).to.equal(BindingStyle.CONSTANT);
+    });
+  });
+
+  describe('toDynamicValue(dynamicValueFn)', () => {
+    it('support a factory', async () => {
+      const b = ctx.bind('msg').toDynamicValue(() => Promise.resolve('hello'));
+      const value: string = await ctx.get('msg');
+      expect(value).to.equal('hello');
+      expect(b.style).to.equal(BindingStyle.DYNAMIC_VALUE);
+    });
+  });
+
+  describe('toClass(cls)', () => {
+    it('support a class', async () => {
+      ctx.bind('msg').toDynamicValue(() => Promise.resolve('world'));
+      const b = ctx.bind('myService').toClass(MyService);
+      expect(b.style).to.equal(BindingStyle.CLASS);
+      const myService: MyService = await ctx.get('myService');
+      expect(myService.getMessage()).to.equal('hello world');
+    });
   });
 
   describe('toProvider(provider)', () => {
     it('binding returns the expected value', async () => {
       ctx.bind('msg').to('hello');
       ctx.bind('provider_key').toProvider(MyProvider);
-      const value: String = await ctx.get('provider_key');
+      const value: string = await ctx.get('provider_key');
       expect(value).to.equal('hello world');
     });
 
     it('can resolve provided value synchronously', () => {
       ctx.bind('msg').to('hello');
       ctx.bind('provider_key').toProvider(MyProvider);
-      const value: String = ctx.getSync('provider_key');
+      const value: string = ctx.getSync('provider_key');
       expect(value).to.equal('hello world');
     });
 
     it('support asynchronous dependencies of provider class', async () => {
       ctx.bind('msg').toDynamicValue(() => Promise.resolve('hello'));
       ctx.bind('provider_key').toProvider(MyProvider);
-      const value: String = await ctx.get('provider_key');
+      const value: string = await ctx.get('provider_key');
       expect(value).to.equal('hello world');
     });
+
+    it('sets the style to PROVIDER', () => {
+      ctx.bind('msg').to('hello');
+      const b = ctx.bind('provider_key').toProvider(MyProvider);
+      expect(b.style).to.equal(BindingStyle.PROVIDER);
+    });
+  });
+
+  describe('toJSON()', () => {
+    it('converts a keyed binding to plain JSON object', () => {
+      const json = binding.toJSON();
+      expect(json).to.eql({
+        key: key,
+        scope: BindingScope.TRANSIENT,
+        tags: [],
+        isLocked: false,
+      });
+    });
+
+    it('converts a binding with more attributes to plain JSON object', () => {
+      const myBinding = new Binding(key, true)
+        .inScope(BindingScope.CONTEXT)
+        .tag('model').to('a');
+      const json = myBinding.toJSON();
+      expect(json).to.eql({
+        key: key,
+        scope: BindingScope.CONTEXT,
+        tags: ['model'],
+        isLocked: true,
+        style: BindingStyle.CONSTANT,
```

Comment:
I am not sure if `style` captures the intent accurately. Would `kind` or `source` be any better? @kjdelisle can you come up with a better suggestions?